File: src/server/middleware/x402.ts
import { Request, Response, NextFunction } from 'express';

interface X402Options {
  price: number;
  token: string;
  gatewayUrl: string;
}

export const x402Middleware = (opts: X402Options) => {
  return (req: Request, res: Response, next: NextFunction) => {
    // Simulate paid check (in prod: validate receipt from gateway)
    if (req.headers['x-receipt']) {
      return next();
    }

    res.status(402).set({
      'Pay': `${opts.gatewayUrl}; amount=${opts.price}; asset=${opts.token}`,
      'Content-Type': 'application/json'
    }).json({
      error: 'Payment Required',
      price: opts.price,
      token: opts.token,
      gateway: opts.gatewayUrl
    });
  };
};

File: gateway/src/middleware/x402.rs
use actix_web::{HttpResponse, Responder, web};
use serde_json::json;

pub fn x402_middleware(
    price: u64,
    token: &str,
    gateway_url: String,
) -> impl Fn(web::ReqData<String>, web::HttpRequest) -> impl std::future::Future<Output = impl Responder> {
    move |merchant_pubkey: web::ReqData<String>, req: web::HttpRequest| {
        async move {
            if req.headers().contains_key("x-receipt") {
                return HttpResponse::Ok().json(json!({
                    "message": "Welcome to private premium content!",
                    "paid_via": "x402 + Elusiv"
                }));
            }

            HttpResponse::PaymentRequired()
                .insert_header(("Pay", format!("{}?amount={}&asset={}", gateway_url, price, token)))
                .json(json!({
                    "error": "Payment Required",
                    "price": price,
                    "token": token,
                    "gateway": gateway_url
                }))
        }
    }
}
